### PR TITLE
✨ EAR 1249 - Add API for locking questionnaires

### PR DESCRIPTION
**Pull request by tbaggaley**
### What is the context of this PR?

[Ticket](https://collaborate2.ons.gov.uk/jira/browse/EAR-1249)

Adds mutation:

`setQuestionnaireLocked` : `{ questionnaireId, locked }` ->
  locks / unlocks questionnaire with given ID based on value of `locked`

Adds field:

`locked` :
  Boolean value determining whether current questionnaire is locked /
  unlocked

Adds migration to add `locked` field to existing questionnaires,
defaulting to `false`.

Prevent modifying questionnaires which are locked by making
`createMutation` check for questionnaire's locked status. Can be
overriden through passing in new `options` object with truthy `ignoreLockStatus`
to `createMutation` - needed so that the locking / unlocking mutation can still operate on
locked questionnaires.

Prevent modifying comments of questionnaires which are locked per AC.

Duplicating a locked questionnaire results in the duplicate being
unlocked per AC.

Have tested using dev tools and all appears to work as intended

### How to review

- Create a questionnaire
- Attempt to lock it using the `setQuestionnaireLocked` mutation. Can use this mutation in the Apollo dev tools:

`mutation Lock($input: SetQuestionnaireLockedInput!) {
  setQuestionnaireLocked(input: $input) {
    id
    locked
  }
}
`

For variables, use:

`{
  "input": {
    "questionnaireId": "<your-questionnaire-id>",
    "locked": true
  }
}`

- No changes should save to the questionnaire, no comments should save or be editable
- Unlock it by running the above mutation with `"locked": false` in your `input` object.
- Should be able to edit as normal

### What to do after everything is green

1. - [ ] Bring this branch up-to-date with Origin/Master
2. - [ ] If there are a lot of commits or some are not helpful to read, squash them down
3. - [ ] Click the **Merge** button on this pull request
4. - [ ] Does this change mean the **Capability examples** questionnaire should be updated?
5. - [ ] Move the Jira ticket for this task into the next stage of the process

### CODE REVIEW

File: eq-author-api/schema/tests/comments.test.js [hunk 1]
```diff
@@ -152,6 +151,22 @@ describe("comments", () => {
...
+          { ...ctx, questionnaire: { ...ctx.questionnaire, locked: true } },
+          {
+            componentId,
+            commentText: "a new comment is created",
+          }
+        )
+      ).rejects.toThrow();
+
+      questionnaire.locked = false;
+    });
```
Comment (MattElcock) [inline eq-author-api/schema/tests/comments.test.js:167]: You could enhance this test by checking that the type of error thrown and message it's thrown with is what is expected.

File: eq-author-api/schema/tests/questionnaire.test.js [hunk 2]
```diff
@@ -237,6 +238,49 @@ describe("questionnaire", () => {
...
+          },
+          ctx
+        );
+
+        expect(
+          updateSurveyId({
+            questionnaireId: ctx.questionnaire.id,
+            surveyId: "1337",
+          })
+        ).rejects.toThrow();
```
Comment (MattElcock) [inline eq-author-api/schema/tests/questionnaire.test.js:280]: You could enhance this test, too, by asserting that the right error is thrown.

File: eq-author-api/schema/tests/questionnaire.test.js [hunk 3]
```diff
@@ -658,6 +702,19 @@ describe("questionnaire", () => {
...
+      await setQuestionnaireLocked(
+        {
+          questionnaireId: ctx.questionnaire.id,
+          locked: true,
+        },
+        ctx
+      );
+
+      expect(deleteQuestionnaire(ctx, ctx.questionnaire.id)).rejects.toThrow();
+    });
```
Comment (MattElcock) [inline eq-author-api/schema/tests/questionnaire.test.js:717]: You could enhance this test, too, by asserting that the right error is thrown.

**Review by MattElcock — DISMISSED** [review]
Looks good to me; nice one, Tom. I've left a few comments but you don't have to do them.

File: eq-author-api/schema/tests/questionnaire.test.js (hunk 3 above)
Comment (tbaggaley) [reply eq-author-api/schema/tests/questionnaire.test.js:717]: Aye, cheers Matt - I did try and have them assert for a `ForbiddenError` but it doesn't actually percolate through to the tests - our (or Apollo's?) wrapper which executes the mutations for the tests always returns a plain old `Error` exception on failure rather the one I actually threw, so not much point checking for that.

File: eq-author-api/schema/resolvers/withPermissions.js [hunk 4]
```diff
@@ -1,24 +1,33 @@
+const { ForbiddenError } = require("apollo-server-express");
+
 const hasWritePermission = (questionnaire, user) =>
   questionnaire.createdBy === user.id ||
   questionnaire.editors.indexOf(user.id) > -1 ||
```
Comment (tjamesmac) [inline eq-author-api/schema/resolvers/withPermissions.js:5]:
```suggestion
  questionnaire.editors.includes(user.id) ||
```
?

**Review by tjamesmac — APPROVED** [review]
That's cool

File: eq-author-api/schema/resolvers/withPermissions.js (hunk 4 above)
Comment (tbaggaley) [reply eq-author-api/schema/resolvers/withPermissions.js:5]: Aye, why not ;-)

**Review by MattElcock — APPROVED** [review]

**Review by tjamesmac — APPROVED** [review]

**Pull request merged**